demo_nodes_cpp/src/topics/listener_content_filter.cpp

```diff
@@ -0,0 +1,67 @@
+// Copyright 2022 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "rclcpp/rclcpp.hpp"
+#include "rclcpp_components/register_node_macro.hpp"
+
+#include "std_msgs/msg/string.hpp"
+
+#include "demo_nodes_cpp/visibility_control.h"
+
+namespace demo_nodes_cpp
+{
+// Create a Listener class that subclasses the generic rclcpp::Node base class.
+// The main function below will instantiate the class as a ROS node.
+class ListenerContentFilter : public rclcpp::Node
+{
+public:
+  DEMO_NODES_CPP_PUBLIC
+  explicit ListenerContentFilter(const rclcpp::NodeOptions & options)
+  : Node("content_filter_listener", options)
+  {
+    // Create a callback function for when messages are received.
+    // Variations of this function also exist using, for example UniquePtr for zero-copy transport.
+    setvbuf(stdout, NULL, _IONBF, BUFSIZ);
+    auto callback =
+      [this](std_msgs::msg::String::ConstSharedPtr msg) -> void
+      {
+        RCLCPP_INFO(this->get_logger(), "I heard: [%s]", msg->data.c_str());
+      };
+    // Create a subscription to the topic which can be matched with one or more compatible ROS
+    // publishers.
+    // Note that not all publishers on the same topic with the same type will be compatible:
+    // they must have compatible Quality of Service policies.
+
+    // Initialize a subscription with a content filter to receive messages that contain a prefix
+    // with 'Hello World: 1', which is similar to the SQL statement `data like 'Hello World: 1%'`.
+    // NOTE: rmw_connextdds does not support `like` operation currently.
```

Comment:
Maybe we could use the `MATCH` feature? Something like `"data MATCH %0"` and `{"'Hello World: \d*[02468]'"}` might match only even numbers?

Reply:
`match` is not standard in the DDS spec.

What about using `data = 'Hello World: 1' or data = 'Hello World: 10'`? What I thought is that it seems not a good experience for the user, so I insist on using the `like`.

Reply:
Match isn't supported? That seems pretty unfortunate. 

This brings up another point. We need to define what can and cannot be used in the filter expressions at the ros level. I don't think it's a good thing to just reference what the underlying middleware supports only.

And I don't think we should use an example that only works with one vendor.

Reply:
So while the "like" example is more compelling, the fact that it is not portable kind of disqualifies it in my mind.

Reply:
To add the limitation at the ros level, I think it's not fair to FastDDS.

I'll try to use `match` instead of `like` at this PR.

Reply:
> match is not standard in the DDS spec.

did you check this `match` does not work? it seems Fast-DDS supports it based on the source code DDSSQLFilter.

Reply:
> did you check this match does not work? 

No, not yet. I was afraid it's not supported in other coming DDS that might not support `match` based on the DDS spec.

Reply:
connextdds(fnmatch) and fastdds(std::regex) use different regular expression. `fnmatch` doesn't know the `{"'Hello World: \\d*[02468]'"}`, which works well in fastdds(std::regex).

Reply:
I'll update content filter just with `"data = %0", {"'Hello World: 10'"}`.

I believe that there will some new issues to ask `why can't I use 'like' for content filter setting?`.

Reply:
To be honest we shouldn't encourage the use of ones outside the standard then (unless we support them at the rcl layer as a fallback) because we cannot use "like" in any meaningful way if it's not required to be supported by the rmw implementation. 

I think your suggestion of matching one exact string is fine. Really the chatter example just isn't well suited to this kind of content filtering it seems.